modules/imgproc/src/sumpixels.simd.hpp

```diff
@@ -114,57 +114,358 @@ struct Integral_SIMD<uchar, int, double>
-                v_store(sum_row + j + v_int32::nlanes, el4h + vx_load(prev_sum_row + j + v_int32::nlanes));
+                    v_store_interleave(row_cache                      , el4l_1, el4l_2, el4l_3);
+                    v_store_interleave(row_cache + v_int32::nlanes * 3, el4h_1, el4h_2, el4h_3);
+                    el4l_1 = vx_load(row_cache                      );
+                    el4l_2 = vx_load(row_cache + v_int32::nlanes    );
+                    el4l_3 = vx_load(row_cache + v_int32::nlanes * 2);
+                    el4h_1 = vx_load(row_cache + v_int32::nlanes * 3);
+                    el4h_2 = vx_load(row_cache + v_int32::nlanes * 4);
+                    el4h_3 = vx_load(row_cache + v_int32::nlanes * 5);
+                    v_store(sum_row + j                      , el4l_1 + vx_load(prev_sum_row + j                      ));
+                    v_store(sum_row + j + v_int32::nlanes    , el4l_2 + vx_load(prev_sum_row + j + v_int32::nlanes    ));
+                    v_store(sum_row + j + v_int32::nlanes * 2, el4l_3 + vx_load(prev_sum_row + j + v_int32::nlanes * 2));
+                    v_store(sum_row + j + v_int32::nlanes * 3, el4h_1 + vx_load(prev_sum_row + j + v_int32::nlanes * 3));
+                    v_store(sum_row + j + v_int32::nlanes * 4, el4h_2 + vx_load(prev_sum_row + j + v_int32::nlanes * 4));
+                    v_store(sum_row + j + v_int32::nlanes * 5, el4h_3 + vx_load(prev_sum_row + j + v_int32::nlanes * 5));
+                }
+
+                for (int v3 = sum_row[j - 1] - prev_sum_row[j - 1],
+                         v2 = sum_row[j - 2] - prev_sum_row[j - 2],
+                         v1 = sum_row[j - 3] - prev_sum_row[j - 3]; j < width; j += 3)
+                {
+                    sum_row[j]     = (v1 += src_row[j])     + prev_sum_row[j];
+                    sum_row[j + 1] = (v2 += src_row[j + 1]) + prev_sum_row[j + 1];
+                    sum_row[j + 2] = (v3 += src_row[j + 2]) + prev_sum_row[j + 2];
+                }
             }
+        }
+        else if (cn == 4)
+        {
+            // the others
+            v_int32 mask = vx_setall_s32((int)0xff);
+            for (int i = 0; i < height; ++i)
+            {
+                const uchar * src_row = src + _srcstep * i;
+                int * prev_sum_row = (int *)((uchar *)sum + _sumstep * i) + cn;
+                int * sum_row = (int *)((uchar *)sum + _sumstep * (i + 1)) + cn;
+
+                sum_row[-1] = sum_row[-2] = sum_row[-3] = sum_row[-4] = 0;
 
-            for (int v = sum_row[j - 1] - prev_sum_row[j - 1]; j < width; ++j)
-                sum_row[j] = (v += src_row[j]) + prev_sum_row[j];
+                v_int32 prev_1 = vx_setzero_s32(), prev_2 = vx_setzero_s32(),
+                        prev_3 = vx_setzero_s32(), prev_4 = vx_setzero_s32();
+                int j = 0;
+                for ( ; j + v_uint16::nlanes * cn <= width; j += v_uint16::nlanes * cn)
```

Comment:
The code looks over-complicated to me. IMO it would be better to process one vector at a time and reduce amount of shifts and additions starting with addition of element quads.